app/code/Magento/MediaGallery/Model/ResourceModel/Keyword/SaveAssetLinks.php

```diff
@@ -56,27 +66,123 @@ public function __construct(
     public function execute(int $assetId, array $keywordIds): void
     {
         try {
-            $values = [];
-            foreach ($keywordIds as $keywordId) {
-                $values[] = [$assetId, $keywordId];
+            $this->deleteAssetKeywords($assetId, $keywordIds);
+            $this->insertAssetKeywords($assetId, $keywordIds);
+        } catch (\Exception $exception) {
+            $this->logger->critical($exception);
+            throw new CouldNotSaveException(
+                __('Could not process asset keyword links'),
+                $exception
+            );
+        }
+    }
+
+    /**
+     * Save new asset keyword links
+     *
+     * @param int $assetId
+     * @param array $keywordIds
+     * @throws CouldNotSaveException
+     */
+    private function insertAssetKeywords(int $assetId, array $keywordIds): void
+    {
+        try {
+            if (!empty($keywordIds)) {
+                $values = [];
+                $keywordsToInsert = array_diff($keywordIds, $this->getCurrentKeywords($assetId));
```

Comment:
What do you think about moving current keywords retrieving to the `execute` method to retrieve them only once.
`insertAssetKeywords` and `deleteAssetKeywords` methods can accept only the needed keywordIds (the result of `array_diff`)